### ISSUE TITLE
std.math.isInf either should support ComptimeFloat or should have a clearer error message

### ISSUE THREAD

**Issue by ssteinbach** [ziglang/zig#22107]
### Zig Version

0.14.0-dev.2340+182cdf74b

### Steps to Reproduce and Observed Behavior

```zig
test {
    _ = std.math.isInf(3.0);
}
```

Produces error:

```
~/zig/0.14.0-dev.2340+182cdf74b/files/lib/std/math/isinf.zig:8:55: error: access of union field 'float' while field 'comptime_float' is active
    const TBits = std.meta.Int(.unsigned, @typeInfo(T).float.bits);
                                          ~~~~~~~~~~~~^~~~~~
~/zig/0.14.0-dev.2340+182cdf74b/files/lib/std/builtin.zig:556:18: note: union declared here
pub const Type = union(enum) {
                 ^~~~~
/var/tmp/inftest.zig:21:23: note: called from here
    _ = std.math.isInf(3.0);
        ~~~~~~~~~~~~~~^~~~~
```

`std.math.ifFinite` produces the same error

### Expected Behavior

Given that `std.math.isNan` supports `ComptimeFloat` arguments, I expected `std.math.isInf` / `std.math.isFinite`.  If there is a reason why it shouldn't support `ComptimeFloat`, then a more helpful @compileError might be helpful.  Just flagging this in case it was missed!  Thanks all!

**Comment by rootbeer** [comment]
I think https://github.com/ziglang/zig/issues/21198 and https://github.com/ziglang/zig/issues/21205 are relevant here.  If I'm reading those correctly, Zig won't support inf/nan in comptime floats (any operations that would generate them at compile time would instead generate a compile error).  I believe this work is in progress (see https://github.com/ziglang/zig/pull/23177 which went in yesterday).

For now, the work-around seems to be a cast: `std.math.isInf(@as(f128, 3.0))`

**Comment by mlugg** [comment]
In light of #21205, it makes sense for `isInf` to *not* support `comptime_float`, in just the same way that it does not support integers/`comptime_int` because while, yes, integer values *aren't* infinity, the query just doesn't really make sense to begin with on those types. #21205 makes `comptime_float` more akin to a "real number" type (albeit with precision around the same as an IEEE-754 binary128 floating-point value).

In terms of the compile error, we typically *avoid* introducing custom `@compileError` logic to `std` in cases like this. As such, I think there is no further work to do here, and will close this issue.

**Comment by silversquirl** [comment]
@mlugg I disagree. generic code may want to call `isInf` without having to check whether it's been passed a `comptime_float`. being able to use `anytype` to accept any floating point type, including `comptime_float`, is useful and I think the standard library should do its best to support this usecase.

in comparison, code that assumes floats will likely not work correctly on integers *anyway*, due to differences in how operations like division work. but comptime_float is close enough to the runtime float types that basically all float code will work fine on it.

**Comment by mlugg** [comment]
I don't think that argument holds up. As an important piece of context, note the existence and acceptance of #23173. With that proposal in mind, we can define three categories of numeric scalar type:

* Integer types (`u8`, `i8`, `u123`, `comptime_int`, ...)
* IEEE-754 floating-point types `float16`...`float128`
* "Real number" types `real16`...`real128`

Crucially, per #21205, `comptime_float` calls into the *last* category, "real number" types, rather than representing an IEEE-754 floating point value.

Operations like division are different between all three of these classes. For integer types, division truncates to an integer; for IEEE-754 types, it follows precise IEEE-754 semantics; for "real number" types, it gives approximately correct numeric results without any specific guarantees on precision etc. I don't think "the results are close enough" is a very good argument for why "real number" types *should* mimic IEEE-754 types (or, more accurately, `std.math` should pretend that they are the same) but integer types should *not*.

A bigger concern is that since `isInf` as a query only makes sense only on IEEE-754 types, using it indicates that you might be expecting IEEE-754 semantics; for instance, perhaps you are using infinities to detect overflowing computations. As such, allowing `isInf` to respond with `false` for "real number" types could potentially mask bugs. My claim here is that using `isInf` etc potentially implies an expectation of specific semantics which types like `comptime_float` do not possess. If the caller does not *require* those semantics, but just needs extra logic to handle inf/nan in the case where those semantics *are* in play, then I think it is better that the caller spell that out explicitly with a type check at the `isInf` call site; that forces them to actually think about what is and isn't correct in the context of "real number" types.

**Comment by ssteinbach** [comment]
Fair enough on not supporting comptime float, but I feel like a more specific compileError for isInf might be helpful so that future users don't assume this a bug or oversight but rather a specific behavior.  It might not be obvious to a novice zig user, for example, what the distinction between a comptime float and a float is in this case.  Either way, thanks for taking a look and thanks for writing up your thoughts!
